homeassistant/components/modbus/validators.py

```diff
@@ -130,20 +162,47 @@ def struct_validator(config: dict[str, Any]) -> dict[str, Any]:
         if not swap_type_validator:
             error = f"{name}: `{CONF_SWAP}:{swap_type}` cannot be combined with `{CONF_DATA_TYPE}: {data_type}`"
             raise vol.Invalid(error)
+    if register_size_bytes == 4 and not validator.register_size_4bytes:
+        error = f"{name}: `{CONF_REGISTER_SIZE_BYTES}: {register_size_bytes}` cannot be specified with `{CONF_DATA_TYPE}: {data_type}`"
+        raise vol.Invalid(error)
+    if register_size_bytes == 8 and not validator.register_size_8bytes:
+        error = f"{name}: `{CONF_REGISTER_SIZE_BYTES}: {register_size_bytes}` cannot be specified with `{CONF_DATA_TYPE}: {data_type}`"
+        raise vol.Invalid(error)
+    if register_size_bytes == 0 or register_size_bytes % 2 != 0:
+        error = f"{name}: Zero or odd numbers are not valid register sizes."
+        raise vol.Invalid(error)
+    if register_size_bytes not in (2, 4, 8):
+        error = f"{name}: `{CONF_REGISTER_SIZE_BYTES}:{register_size_bytes}` is out of range. Accepted values are 2, 4 or 8 bytes."
+        raise vol.Invalid(error)
     if config[CONF_DATA_TYPE] == DataType.CUSTOM:
         try:
             size = struct.calcsize(structure)
         except struct.error as err:
             raise vol.Invalid(
                 f"{name}: error in structure format --> {str(err)}"
             ) from err
-        bytecount = count * 2
+        bytecount = count * register_size_bytes
         if bytecount != size:
             raise vol.Invalid(
                 f"{name}: Size of structure is {size} bytes but `{CONF_COUNT}: {count}` is {bytecount} bytes"
             )
     else:
-        if data_type != DataType.STRING:
+        if data_type not in (DataType.STRING, DataType.INT8, DataType.UINT8):
+            if (
+                DEFAULT_STRUCT_FORMAT[data_type].register_count
+                * 2
+                % register_size_bytes
+                != 0
+            ):
+                raise vol.Invalid(
+                    "Combination of data type and register size does generate an posivite integer count."
+                )
+            config[CONF_COUNT] = int(
+                DEFAULT_STRUCT_FORMAT[data_type].register_count
```

Comment:
Please use count directly from the table. Calculations like this are hard to maintain over time, because you forget the details.

Reply:
I've tried to add the different register counts in the table indexed by the register size. Not totally happy with having .register_count.count_for_register_size[ <register_size>], would rather do .register_count[ <register_size> ] though tried to follow the convention in the table with the namedtuples wasnt sure if adding a dictionary directly would be appropriate.